### ISSUE TITLE
Issue reproducing same results as that of the first training

### ISSUE THREAD

**Issue by qinghezeng** [mahmoodlab/CLAM#62]
Hi fedshyvana, do you have an idea why running a same training command directly in terminal, will lead to different results compared to using a bash script?

For example, at the beginning of the training using a bash script:

> batch 4, loss: 0.5872, instance_loss: 1.3377, weighted_loss: 0.8123, inst_p_acc: 0.0000, inst_n_acc: 0.3750, label: 0, bag_size: 10408

and directly executing in a terminal:

> batch 4, loss: 0.5863, instance_loss: 1.3010, weighted_loss: 0.8007, inst_p_acc: 0.5000, inst_n_acc: 0.6250, label: 0, bag_size: 10408 

I verified this issue could be reproduced on my computer. And I didn't explicitly set a seed so I think seed 1 was used by default.

Question 2: Also I wonder why in terminal, re-executing a training command can lead to slightly different results. 

For example, the training became slightly different after some epochs,

> batch 209, loss: 0.4016, instance_loss: 0.0345, weighted_loss: 0.2914, inst_p_acc: 1.0000, inst_n_acc: 1.0000, label: 0, bag_size: 2985

> batch 209, loss: 0.4016, instance_loss: 0.0344, weighted_loss: 0.2914, inst_p_acc: 1.0000, inst_n_acc: 1.0000, label: 0, bag_size: 2985

How could I reproduce the exactly same results as that of the first training? Any help or idea would be appreciated,  thank you!

**Comment by fedshyvana** [comment]
i don't have the answer regarding your first answer sadly, but regarding question 2, this is the official statement from pytorch regarding reproducibility:
"Completely reproducible results are not guaranteed across PyTorch releases, individual commits, or different platforms. Furthermore, results may not be reproducible between CPU and GPU executions, even when using identical seeds."

**Comment by fedshyvana** [comment]
https://pytorch.org/docs/stable/notes/randomness.html

**Comment by qinghezeng** [comment]
Ok thanks for your quick answer!
